### PR TITLE
api, core: support zero copy into protobuf

### PR DESCRIPTION
**Pull request by voidzcy**
A variant of #7330, taking the alternative codepath in protobuf.

Major changes of this PR: 
   - Added a `HasByteBuffer` API that allows the marshaller to access the backing ByteBuffers directly (cherry-picked from #7330).
   - Added a `Detachable` API that allows the application to take over the ownership of underlying buffers and close them later.

The new approach is to let the application implement a marshaller that passes ByteBuffers as immutable to protobuf:
   - Wrap ByteBuffers to ByteString with [UnsafeByteOperations](https://github.com/protocolbuffers/protobuf/blob/bc45f92262a8e0e6e1ab7302c72a380eb0346f8e/java/core/src/main/java/com/google/protobuf/UnsafeByteOperations.java#L97)
   - [Concatenate](https://github.com/protocolbuffers/protobuf/blob/bc45f92262a8e0e6e1ab7302c72a380eb0346f8e/java/core/src/main/java/com/google/protobuf/ByteString.java#L588) them as a RopeByteString
   - [Turn it to CodedInputStream](https://github.com/protocolbuffers/protobuf/blob/bc45f92262a8e0e6e1ab7302c72a380eb0346f8e/java/core/src/main/java/com/google/protobuf/RopeByteString.java#L605)

This requires buffers to be alive all the time through the proto messages used by the application (compared to the normal codepath that makes a copy of the bytes when parsed to proto messages and recycle buffers right after). The application is responsible for managing the lifetime of underlying buffers. An example marshaller that enables zero-copy codepath can be something like the following:

```java
  class ZeroCopyMarshaller<T extends Message> implements PrototypeMarshaller<T> {
    private IdentityHashMap<T, InputStream> unclosedStreams = new IdentityHashMap<>();
    private final Parser<T> parser;

    ZeroCopyMarshaller(T defaultInstance) {
      parser = (Parser<T>) defaultInstance.getParserForType();
    }

    // ... ...

    @Override
    public T parse(InputStream stream) {
      CodedInputStream cis = null;
      boolean streamDetached = false;
      if (stream instanceof Detachable) {
        stream = ((Detachable) stream).detach();
        streamDetached = true;
      }
      try {
        if (stream instanceof KnownLength) {
          int size = stream.available();
          if (stream instanceof HasByteBuffer
              && ((HasByteBuffer) stream).byteBufferSupported()) {  // fastest path, zero copy
            List<ByteString> byteStrings = new ArrayList<>();
            while (stream.available() != 0) {
              ByteBuffer buffer = ((HasByteBuffer) stream).getByteBuffer();
              byteStrings.add(UnsafeByteOperations.unsafeWrap(buffer));
              stream.skip(buffer.remaining());
            }
            cis = ByteString.copyFrom(byteStrings).newCodedInput();
          } else {
            // slightly slower path, copy into a byte array
          }
        }
      } catch (IOException e) {
        throw new RuntimeException(e);
      }
      if (cis == null) {  // slowest path, via InputStream read
        cis = CodedInputStream.newInstance(stream);
      }

      // ... ...

      T message;
      try {
        message = parseFrom(cis);
      } catch (InvalidProtocolBufferException ipbe) {
        throw Status.INTERNAL.withDescription("Invalid protobuf byte sequence").withCause(ipbe).asRuntimeException();
      }
      if (streamDetached) {
        unclosedStreams.put(message, stream);
      }
      return message;
    }

    private T parseFrom(CodedInputStream stream) throws InvalidProtocolBufferException {
      T message = parser.parseFrom(stream);
      try {
        stream.checkLastTagWas(0);
        return message;
      } catch (InvalidProtocolBufferException e) {
        e.setUnfinishedMessage(message);
        throw e;
      }
    }

    // Application calls this method when no longer need the message.
    public void releaseMessage(T message) {
      InputStream stream = unclosedStreams.get(message);
      if (stream != null) {
        try {
          stream.close();
        } catch (IOException e) {
          throw new RuntimeException(e);
        }
        unclosedStreams.remove(message);
      }
    }
  }
```

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java [hunk 1]
```diff
@@ -147,46 +161,79 @@ public void readBytes(OutputStream dest, int length) throws IOException {
...
+          dest.addBuffer(buffer.readBytes(length));
+          return 0;
+        }
+      };
+
   @Override
   public ReadableBuffer readBytes(int length) {
-    if (length <= 0) {
-      return ReadableBuffers.empty();
+    final CompositeReadableBuffer newBuffer = new CompositeReadableBuffer();
```
Comment (voidzcy) [inline core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java:171]: Unfortunately, this reverts the optimization made in #7375. We cannot directly take the buffer instance from the source composite, as it should not retain the old mark position after the read.

[the remark this answers is not in the document]
Comment (voidzcy) [reply core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java:171]: Oops, I commented in https://github.com/grpc/grpc-java/pull/8102#discussion_r618687981

[the remark this answers is not in the document]
Comment (njhill) [reply core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java:171]: @voidzcy returning the buffer directly is only one part of the optimization. You can still avoid allocating a composite buffer since `readBytes(length)` already allocates a new buffer with its own marker.

How about this:
```java
    ReadableBuffer newBuffer = null;
    CompositeReadableBuffer newComposite = null;
    do {
      ReadableBuffer buffer = buffers.peek();
      int readable = buffer.readableBytes();
      ReadableBuffer readBuffer;
      if (readable > length) {
        readBuffer = buffer.readBytes(length);
        length = 0;
      } else {
        if (marked) {
          readBuffer = buffer.readBytes(readable);
          advanceBufferIfNecessary(); // it _is_ necessary, so this will do couple of redundant checks
        } else {
          readBuffer = buffers.poll();
        }
        length -= readable;
      }
      if (newBuffer == null) {
        newBuffer = readBuffer;
      } else {
        if (newComposite == null) {
          newComposite =
              new CompositeReadableBuffer(length == 0 ? 2 : Math.min(buffers.size() + 2, 16));
          newComposite.addBuffer(newBuffer);
          newBuffer = newComposite;
        }
        newComposite.addBuffer(readBuffer);
      }
    } while (length > 0);
    return newBuffer;
```

[the remark this answers is not in the document]
Comment (voidzcy) [reply core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java]: Restore the change. I previously thought that was too verbose compared to the amount of optimizations it can offer :P

[the remark this answers is not in the document]
Comment (njhill) [reply core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java]: Thanks @voidzcy, agree that the other approach is much more elegant.

File: core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java [hunk 2]
```diff
@@ -33,15 +35,19 @@
...
   public CompositeReadableBuffer(int initialCapacity) {
-    buffers = new ArrayDeque<>(initialCapacity);
+    readableBuffers = new ArrayDeque<>(initialCapacity);
+    rewindableBuffers = new ArrayDeque<>(initialCapacity);
   }
 
   public CompositeReadableBuffer() {
-    buffers = new ArrayDeque<>();
+    readableBuffers = new ArrayDeque<>();
+    rewindableBuffers = new ArrayDeque<>();
```
Comment (njhill) [inline core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java:50]: This could be initialized to a constant empty queue, and replaced with a new ArrayDeque in the one place that it is appended to (in `advanceBuffer()`). It would save couple more allocations in cases that mark/reset isn't used.

Comment (voidzcy) [reply core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java]: Alright, changed to instantiate `rewindableBuffers` lazily.

File: api/src/main/java/io/grpc/HasByteBuffer.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ * interface from an {@link java.io.InputStream} to {@link ByteBuffer}s, without copying the
+ * content to a byte array and read from it.
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/7387")
+public interface HasByteBuffer {
+
+  /**
+   * Indicates whether or not {@link #getByteBuffer} operation is supported.
+   */
+  boolean getByteBufferSupported();
```
Comment (ejona86) [inline api/src/main/java/io/grpc/HasByteBuffer.java:37]: Did we discuss `isByteBufferSupported` as the name here? That would be the bean-style name. The InputStream-style name would be `byteBufferSupported`.

File: api/src/main/java/io/grpc/Retainable.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+/**
+ * A <i>Retainable</i> is a resource that can be retained, preventing it being closed too early.
```
Comment (ejona86) [inline api/src/main/java/io/grpc/Retainable.java:20]: I'm not wild about how this makes close() a noop and how there is no feedback when the methods are used incorrectly. I think we need an interface that can fail the retain or a way to learn it isn't supported.

I think I was envisioning more a `duplicate()`/`clone()` style of operation that would return a new InputStream. That would keep close() semantics clear. But I'd have to dig deeper to see if that is a pain to implement.

If we want to keep the general approach here, I think we'd want to keep close() doing something even when retained. That could be done by having the semantics be that the inputstream is alive until release()+close(), but in either order. Close() would keep its normal idemponent behavior. But we'd need retain/release to throw instead of being noops if already closed/retained. With this approach it is really weird that you can continue read()ing successfully after close(); seems asking for weird behavior, although I guess this will be used rarely enough to not matter much.

File: api/src/main/java/io/grpc/HasByteBuffer.java (hunk 3 above)
Comment (voidzcy) [reply api/src/main/java/io/grpc/HasByteBuffer.java]: `getByteBufferSupported` indeed looks like an InputStream-style name, doesn't it? Similar to `markSupported()`/`mark()`, here we have `getByteBufferSupported()`/`getByteBuffer()`.

Anyway, I am fine with `byteBufferSupported`.

File: api/src/main/java/io/grpc/Retainable.java (hunk 4 above)
Comment (voidzcy) [reply api/src/main/java/io/grpc/Retainable.java]: Updated to a `duplicate()`/`clone()` style API. See `Detachable`.

In the new API:
- If the Detachable has already been detached, detaching it again would throw.
- Continue reading the detached InputStream behaves as if the InputStream was exhausted; reading the new InputStream behaves as if continuing reading the old InputStream.
- Mark of the original InputStream is discarded after being detached, as the underlying bytes are gone. The new InputStream preserves the mark transferred from the old InputStream.

**Comment by voidzcy** [comment]
@ejona86 Updated. PTAL and let me know if anything I am misunderstanding. Thanks you!

File: api/src/main/java/io/grpc/Detachable.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ * states of detached resources and is no longer consumable as if the resources are exhausted.
+ */
+@ExperimentalApi("TODO")
+public interface Detachable {
+
+  /**
+   * Fork a new instance with underlying resources detached from this instance and
+   * transferred to the new instance.
+   *
+   * @throws IllegalStateException if the underlying resources have already been detached.
```
Comment (ejona86) [inline api/src/main/java/io/grpc/Detachable.java:34]: How is the caller supposed to avoid this exception? What if detachable isn't supported?

I suspect we'll want to allow detach to be called multiple times, but the later times obvious don't return anything "real". We could define it as throwing when `available() == 0`, but that seems an abuse of available() without KnownLength. The answers here may be more obvious if this interface becomes tied to InputStream.

File: api/src/main/java/io/grpc/Detachable.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+   */
+  public Detachable detach();
```
Comment (ejona86) [inline api/src/main/java/io/grpc/Detachable.java:36]: Hmm... I see how the typing is awkward. It is probably more useful for us to return InputStream here.

**Review by ejona86 — COMMENTED** [review]
The comments for Detachable are probably best discussed during API review and see what people think. I think this looks pretty good. Main thing is I'd hope is we don't need all the `if (detached)` checks in the implementation.

File: api/src/main/java/io/grpc/Detachable.java (hunk 6 above)
Comment (ejona86) [reply api/src/main/java/io/grpc/Detachable.java:36]: > Returning InputStream is definitely great for typing. But doesn't it limit this API to InputStream only? Is that desirable for us/future? Would need to update the overall doc/specification if this only applies to InputStream.

Yes, maybe, yes :smile:. Yeah, it'd be nice if generics could manage this, but since this is a tagging interface the caller will be casting which makes generics useless.

I'm not all that concerned about making it generic and usable with any types; I'm fine with focusing on InputStream. My comment about returning InputStream, is simply that Detachable by itself is a useless type to return; you can't do anything with it. That said, maybe you are right that it conveys "you get the same type of thing" a bit more. In either case, we might want to focus the documentation on InputStream more so that it is more clear what its purpose is and how it is used.

File: core/src/main/java/io/grpc/internal/ReadableBuffers.java [hunk 7]
```diff
@@ -329,8 +379,65 @@ public int read(byte[] dest, int destOffset, int length) throws IOException {
...
+      }
+      return buffer.getByteBuffer();
+    }
+
+    @Override
+    public Detachable detach() {
+      if (detached) {
+        throw new IllegalStateException("already detached");
+      }
+      detached = true;
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply core/src/main/java/io/grpc/internal/ReadableBuffers.java]: > Although it is sort of strange, an even easier approach would be to replace `buffer` with `buffer.readBytes(0)` without saving any info about markSupported/etc.

Great idea. Updated with it.

File: api/src/main/java/io/grpc/Detachable.java (hunk 6 above)
Comment (voidzcy) [reply api/src/main/java/io/grpc/Detachable.java]: Updated.

File: api/src/main/java/io/grpc/Detachable.java (hunk 5 above)
Comment (voidzcy) [reply api/src/main/java/io/grpc/Detachable.java]: Updated.

**Comment by voidzcy** [comment]
Updated descriptions in this PR and ExperimentalApi links in the code. Please let me know if any doc needs improvement. Thanks for the review!

File: api/src/main/java/io/grpc/Detachable.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import java.io.InputStream;
+
+/**
+ * A <i>Detachable</i> encapsulates some readable data source that can be detached and transferred
```
Comment (ejona86) [inline api/src/main/java/io/grpc/Detachable.java:22]: I think we need to "throw the user a bone" here and give them a better hint why this exists. Let's say something about how the detached instance can outlive the original instance. And how it may be useful in a Marshaller to perform delayed deserialization or when combined with HasByteBuffer.

An really, let's rephrase the documentation to describe this as "An extension of InputStream." I think it will make it much more clear what is going on.

**Review by ejona86 — APPROVED** [review]

File: api/src/main/java/io/grpc/Detachable.java (hunk 8 above)
Comment (voidzcy) [reply api/src/main/java/io/grpc/Detachable.java]: Updated.

**Review by ejona86 — APPROVED** [review]

**Comment by veblush** [comment]
Thank you for the PR! My early benchmark shows that this can save ~20% of cpu time when used in the GCS benchmark client. ([link](https://github.com/GoogleCloudPlatform/grpc-gcp-java/pull/77#issuecomment-843627151))
